File: sando_py/sando_py/nodes/odom_to_global_state.py
"""Odometry (local) -> global-frame State + PoseStamped — port of odom_to_global_state.cpp.

Looks up TF: world -> <veh_name>/init_pose, caches it, then transforms every
incoming Odometry message into the global "world" frame.
"""
from __future__ import annotations

import math
from typing import Optional

import numpy as np

import rclpy
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.time import Time

from nav_msgs.msg import Odometry
from geometry_msgs.msg import PoseStamped

import tf2_ros
from tf2_ros import Buffer, TransformListener


def _state_msg():
    try:
        from dynus_interfaces.msg import State
        return State
    except ImportError:
        return None


def _quat_mul(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """Hamilton quaternion product (xyzw)."""
    ax, ay, az, aw = a
    bx, by, bz, bw = b
    return np.array([
        aw * bx + ax * bw + ay * bz - az * by,
        aw * by - ax * bz + ay * bw + az * bx,
        aw * bz + ax * by - ay * bx + az * bw,
        aw * bw - ax * bx - ay * by - az * bz,
    ])


def _quat_to_rot(q: np.ndarray) -> np.ndarray:
    qx, qy, qz, qw = q
    return np.array([
        [1 - 2 * (qy * qy + qz * qz), 2 * (qx * qy - qz * qw), 2 * (qx * qz + qy * qw)],
        [2 * (qx * qy + qz * qw), 1 - 2 * (qx * qx + qz * qz), 2 * (qy * qz - qx * qw)],
        [2 * (qx * qz - qy * qw), 2 * (qy * qz + qx * qw), 1 - 2 * (qx * qx + qy * qy)],
    ])


class OdomToGlobalState(Node):
    def __init__(self):
        super().__init__("odom_to_global_state")

        ns = self.get_namespace() or "/"
        self.veh_name = ns.rstrip("/").split("/")[-1] or "PX04"

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        StateMsg = _state_msg()
        if StateMsg is None:
            raise RuntimeError("dynus_interfaces not available")
        self.pub_state = self.create_publisher(StateMsg, "state", 10)
        self.pub_pose = self.create_publisher(PoseStamped, "global_pose", 10)
        self.sub_odom = self.create_subscription(Odometry, "odom", self._odom_cb, 10)
        self.tf_timer = self.create_timer(0.1, self._poll_tf)

        self.tf_acquired = False
        self.t_init = np.zeros(3)
        self.q_init = np.array([0.0, 0.0, 0.0, 1.0])  # xyzw
        self.R_init = np.eye(3)
        self.get_logger().info(f"Waiting for TF: world -> {self.veh_name}/init_pose")

    def _poll_tf(self) -> None:
        target = "world"
        source = f"{self.veh_name}/init_pose"
        if not self.tf_buffer.can_transform(target, source, Time()):
            return
        try:
            tf = self.tf_buffer.lookup_transform(target, source, Time())
        except tf2_ros.TransformException as ex:
            self.get_logger().warn(f"TF lookup failed: {ex}", throttle_duration_sec=2.0)
            return
        self.t_init = np.array([tf.transform.translation.x,
                                tf.transform.translation.y,
                                tf.transform.translation.z])
        self.q_init = np.array([tf.transform.rotation.x,
                                tf.transform.rotation.y,
                                tf.transform.rotation.z,
                                tf.transform.rotation.w])
        self.R_init = _quat_to_rot(self.q_init)
        self.tf_acquired = True
        self.tf_timer.cancel()
        yaw = math.atan2(self.R_init[1, 0], self.R_init[0, 0])
        self.get_logger().info(
            f"TF acquired: t=[{self.t_init[0]:.3f}, {self.t_init[1]:.3f}, "
            f"{self.t_init[2]:.3f}], yaw={yaw:.3f} rad"
        )

    def _odom_cb(self, msg: Odometry) -> None:
        if not self.tf_acquired:
            return
        local_pos = np.array([msg.pose.pose.position.x,
                              msg.pose.pose.position.y,
                              msg.pose.pose.position.z])
        local_vel = np.array([msg.twist.twist.linear.x,
                              msg.twist.twist.linear.y,
                              msg.twist.twist.linear.z])
        local_quat = np.array([msg.pose.pose.orientation.x,
                               msg.pose.pose.orientation.y,
                               msg.pose.pose.orientation.z,
                               msg.pose.pose.orientation.w])
        global_pos = self.R_init @ local_pos + self.t_init
        global_vel = self.R_init @ local_vel
        global_quat = _quat_mul(self.q_init, local_quat)
        n = float(np.linalg.norm(global_quat))
        if n > 1e-9:
            global_quat = global_quat / n

        StateMsg = _state_msg()
        if StateMsg is None:
            return
        s = StateMsg()
        s.header.stamp = msg.header.stamp
        s.header.frame_id = "world"
        s.pos.x, s.pos.y, s.pos.z = float(global_pos[0]), float(global_pos[1]), float(global_pos[2])
        s.vel.x, s.vel.y, s.vel.z = float(global_vel[0]), float(global_vel[1]), float(global_vel[2])
        s.quat.x = float(global_quat[0]); s.quat.y = float(global_quat[1])
        s.quat.z = float(global_quat[2]); s.quat.w = float(global_quat[3])
        self.pub_state.publish(s)

        p = PoseStamped()
        p.header.stamp = msg.header.stamp
        p.header.frame_id = "world"
        p.pose.position.x, p.pose.position.y, p.pose.position.z = (
            float(global_pos[0]), float(global_pos[1]), float(global_pos[2]),
        )
        p.pose.orientation.x = float(global_quat[0])
        p.pose.orientation.y = float(global_quat[1])
        p.pose.orientation.z = float(global_quat[2])
        p.pose.orientation.w = float(global_quat[3])
        self.pub_pose.publish(p)


def main(args=None):
    rclpy.init(args=args)
    node = OdomToGlobalState()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: cpp/ros2/src/sando_cpp/src/odom_to_global_state.cpp
// odom_to_global_state.cpp — C++ ROS2 port of sando_py/nodes/odom_to_global_state.py.
// Looks up TF world -> <veh>/init_pose once, caches (t_init, q_init, R_init), then maps every
// local Odometry into the global `world` frame:  p_g = R_init*p_l + t_init,  v_g = R_init*v_l,
// q_g = q_init (x) q_l (normalized). Publishes State (`state`) + PoseStamped (`global_pose`).
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <dynus_interfaces/msg/state.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <Eigen/Dense>
#include <cmath>
#include <chrono>
using namespace std::chrono_literals;

static Eigen::Vector4d quat_mul(const Eigen::Vector4d& a, const Eigen::Vector4d& b) {
  double ax = a[0], ay = a[1], az = a[2], aw = a[3];
  double bx = b[0], by = b[1], bz = b[2], bw = b[3];
  return {aw * bx + ax * bw + ay * bz - az * by,
          aw * by - ax * bz + ay * bw + az * bx,
          aw * bz + ax * by - ay * bx + az * bw,
          aw * bw - ax * bx - ay * by - az * bz};
}
static Eigen::Matrix3d quat_to_rot(const Eigen::Vector4d& q) {
  double qx = q[0], qy = q[1], qz = q[2], qw = q[3];
  Eigen::Matrix3d R;
  R << 1 - 2 * (qy * qy + qz * qz), 2 * (qx * qy - qz * qw), 2 * (qx * qz + qy * qw),
       2 * (qx * qy + qz * qw), 1 - 2 * (qx * qx + qz * qz), 2 * (qy * qz - qx * qw),
       2 * (qx * qz - qy * qw), 2 * (qy * qz + qx * qw), 1 - 2 * (qx * qx + qy * qy);
  return R;
}

class OdomToGlobalState : public rclcpp::Node {
 public:
  OdomToGlobalState() : rclcpp::Node("odom_to_global_state") {
    std::string ns = get_namespace(); if (ns.empty()) ns = "/";
    while (!ns.empty() && ns.back() == '/') ns.pop_back();
    auto p = ns.find_last_of('/');
    veh_name_ = (p == std::string::npos) ? ns : ns.substr(p + 1);
    if (veh_name_.empty()) veh_name_ = "PX04";

    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(get_clock());
    tf_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf_buffer_);
    pub_state_ = create_publisher<dynus_interfaces::msg::State>("state", 10);
    pub_pose_ = create_publisher<geometry_msgs::msg::PoseStamped>("global_pose", 10);
    sub_odom_ = create_subscription<nav_msgs::msg::Odometry>(
        "odom", 10, std::bind(&OdomToGlobalState::odom_cb, this, std::placeholders::_1));
    tf_timer_ = create_wall_timer(100ms, std::bind(&OdomToGlobalState::poll_tf, this));
    RCLCPP_INFO(get_logger(), "Waiting for TF: world -> %s/init_pose", veh_name_.c_str());
  }

 private:
  void poll_tf() {
    std::string source = veh_name_ + "/init_pose";
    if (!tf_buffer_->canTransform("world", source, tf2::TimePointZero)) return;
    geometry_msgs::msg::TransformStamped tf;
    try {
      tf = tf_buffer_->lookupTransform("world", source, tf2::TimePointZero);
    } catch (const tf2::TransformException& ex) {
      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 2000, "TF lookup failed: %s", ex.what());
      return;
    }
    t_init_ = Eigen::Vector3d(tf.transform.translation.x, tf.transform.translation.y,
                              tf.transform.translation.z);
    q_init_ = Eigen::Vector4d(tf.transform.rotation.x, tf.transform.rotation.y,
                              tf.transform.rotation.z, tf.transform.rotation.w);
    R_init_ = quat_to_rot(q_init_);
    tf_acquired_ = true;
    tf_timer_->cancel();
    double yaw = std::atan2(R_init_(1, 0), R_init_(0, 0));
    RCLCPP_INFO(get_logger(), "TF acquired: t=[%.3f, %.3f, %.3f], yaw=%.3f rad",
                t_init_[0], t_init_[1], t_init_[2], yaw);
  }

  void odom_cb(const nav_msgs::msg::Odometry::SharedPtr msg) {
    if (!tf_acquired_) return;
    Eigen::Vector3d lp(msg->pose.pose.position.x, msg->pose.pose.position.y, msg->pose.pose.position.z);
    Eigen::Vector3d lv(msg->twist.twist.linear.x, msg->twist.twist.linear.y, msg->twist.twist.linear.z);
    Eigen::Vector4d lq(msg->pose.pose.orientation.x, msg->pose.pose.orientation.y,
                       msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);
    Eigen::Vector3d gp = R_init_ * lp + t_init_;
    Eigen::Vector3d gv = R_init_ * lv;
    Eigen::Vector4d gq = quat_mul(q_init_, lq);
    double n = gq.norm();
    if (n > 1e-9) gq /= n;

    dynus_interfaces::msg::State s;
    s.header.stamp = msg->header.stamp; s.header.frame_id = "world";
    s.pos.x = gp[0]; s.pos.y = gp[1]; s.pos.z = gp[2];
    s.vel.x = gv[0]; s.vel.y = gv[1]; s.vel.z = gv[2];
    s.quat.x = gq[0]; s.quat.y = gq[1]; s.quat.z = gq[2]; s.quat.w = gq[3];
    pub_state_->publish(s);

    geometry_msgs::msg::PoseStamped ps;
    ps.header.stamp = msg->header.stamp; ps.header.frame_id = "world";
    ps.pose.position.x = gp[0]; ps.pose.position.y = gp[1]; ps.pose.position.z = gp[2];
    ps.pose.orientation.x = gq[0]; ps.pose.orientation.y = gq[1];
    ps.pose.orientation.z = gq[2]; ps.pose.orientation.w = gq[3];
    pub_pose_->publish(ps);
  }

  std::string veh_name_;
  bool tf_acquired_ = false;
  Eigen::Vector3d t_init_ = Eigen::Vector3d::Zero();
  Eigen::Vector4d q_init_ = Eigen::Vector4d(0, 0, 0, 1);
  Eigen::Matrix3d R_init_ = Eigen::Matrix3d::Identity();
  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
  std::unique_ptr<tf2_ros::TransformListener> tf_listener_;
  rclcpp::Publisher<dynus_interfaces::msg::State>::SharedPtr pub_state_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pub_pose_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
  rclcpp::TimerBase::SharedPtr tf_timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<OdomToGlobalState>());
  rclcpp::shutdown();
  return 0;
}
